spec/rubocop/cop/style/array_and_hash_compact_spec.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::Style::ArrayAndHashCompact do
+  subject(:cop) { described_class.new }
+
+  it 'registers an offense and corrects when using `reject` on array to reject nils' do
+    expect_offense(<<~RUBY)
+      array.reject { |e| e.nil? }
+            ^^^^^^^^^^^^^^^^^^^^^ Use `compact` instead of `reject { |e| e.nil? }`.
+      array.reject! { |e| e.nil? }
+            ^^^^^^^^^^^^^^^^^^^^^^ Use `compact!` instead of `reject! { |e| e.nil? }`.
+    RUBY
+
+    expect_correction(<<~RUBY)
+      array.compact
+      array.compact!
+    RUBY
+  end
+
+  it 'registers an offense and corrects when using `reject` on hash to reject nils' do
+    expect_offense(<<~RUBY)
+      hash.reject { |k, v| v.nil? }
+           ^^^^^^^^^^^^^^^^^^^^^^^^ Use `compact` instead of `reject { |k, v| v.nil? }`.
+      hash.reject! { |k, v| v.nil? }
+           ^^^^^^^^^^^^^^^^^^^^^^^^^ Use `compact!` instead of `reject! { |k, v| v.nil? }`.
+    RUBY
+
+    expect_correction(<<~RUBY)
+      hash.compact
+      hash.compact!
+    RUBY
+  end
+
+  it 'registers an offense and corrects when using `select/select!` to reject nils' do
+    expect_offense(<<~RUBY)
+      array.select { |e| !e.nil? }
+            ^^^^^^^^^^^^^^^^^^^^^^ Use `compact` instead of `select { |e| !e.nil? }`.
+      hash.select! { |k, v| !v.nil? }
+           ^^^^^^^^^^^^^^^^^^^^^^^^^^ Use `compact!` instead of `select! { |k, v| !v.nil? }`.
```

Comment:
There should also be negative cases - selecting nil and rejecting non-nils. Although I cannot image where one would write something like that, ideally the cop should not flag those.